docs/tutorials/demo_best_practices.py

```diff
@@ -0,0 +1,281 @@
+)
+
+# Test the function
+df_test = add_rain_indicators(df_test)
+assert (
+    "is_raining" in df_test.columns and "is_raining_in_1_day" in df_test.columns
+), "Both indicator columns should be present"
+assert all(
+    df_test.loc[24:, "is_raining"]
+), "All values should be True in 'is_raining' for the second day"
+assert all(
+    df_test.loc[:23, "is_raining_in_1_day"]
+), "All values should be True in 'is_raining_in_1_day' for the first day"
+assert df_test.isna().sum().sum() == 0, "There should be no NaN values in the DataFrame"
+
+# %% [markdown]
+"""
+Encapsulating data processing steps in unit tested functions (which are also made available to the whole team or organisation) saves some work by reusing our code.
+Additionally, the tests increase our confidence in the data quality and serve as additional context to infer the purpose of the function should we or someone else come back at a later time.
+
+We can now apply the functions to process the data.
+"""
+
+# %%
+json_data = load_json_data(outfile)
+df = create_dataframe_from_json(json_data)
+df = convert_time_column_to_datetime(df)
+df = add_rain_indicators(df)
+
+df.head(5)
+
+# %% [markdown]
+"""
+## Use commits as checkpoints and merge branches for atomic changes
+
+Now we can commit the updated data to the transform-raw-data branch in lakeFS repository.
+We write a descriptive commit message.
+"""
+# %%
+with fs.transaction as tx:
+    df.to_csv(f"lakefs://{REPO_NAME}/main/weather_2010.csv")
+    commit = tx.commit(repository=REPO_NAME, branch="main", message="Preprocess 2010 data")
+print(commit)
+
+# %% [markdown]
+"""
+We see that the commit has a unique id, the commit SHA which we can log to identify this particular state of the data.
+
+As the data looks good, we can also merge the branch back into `main`.
+"""
+
+# %%
+with fs.transaction as tx:
+    tx.merge(repository=REPO_NAME, source_ref=NEW_BRANCH_NAME, into="main")
+
+# %% [markdown]
+"""
+We will now start to develop our ML model. We recommend checking out a new branch for every set of experiments.
+Here, we will conduct the train test split and further experiment specific modifications, if applicable.
+"""
+
+# %%
+TRAINING_BRANCH = "experiment-1"
+with fs.transaction as tx:
+    tx.create_branch(repository=REPO_NAME, name=TRAINING_BRANCH, source="main")
+
+df = pd.read_csv(f"lakefs://{REPO_NAME}/{TRAINING_BRANCH}/weather_2010.csv")
+model_data = df.drop("time", axis=1)
+train, test = sklearn.model_selection.train_test_split(model_data, random_state=7)
+
+# %% [markdown]
+"""
+## Descriptive tags for human readability and unique SHAs for identification
+
+Since the train test split of the new branch is something we expect to address quite often in development, we will also add a human-readable tag to it.
+In the code below we directly pass the commit as the reference. If we would pass a branch, the tag would point to the current HEAD (i.e. the latest) commit on the branch.
```

Comment:
No conditional in the "if"-clause (see <https://dictionary.cambridge.org/grammar/british-grammar/conditionals-if>):
```suggestion
In the code below we directly pass the commit as the reference. If we pass a branch, the tag will point to the current HEAD (i.e., the latest) commit on the branch.
```